src/aliceVision/depthMap/cuda/commonStructures.hpp

```diff
@@ -131,441 +155,748 @@ CudaSize<Dim> operator-(const CudaSize<Dim> &lhs, const CudaSize<Dim> &rhs) {
+{
+    dst.copyFrom( src );
 }
 
 template<class Type> void copy(CudaHostMemoryHeap<Type,1>& _dst, const CudaDeviceMemory<Type>& _src)
 {
   cudaMemcpyKind kind = cudaMemcpyDeviceToHost;
-  cudaError_t err = cudaMemcpy(_dst.getBuffer(), _src.getBuffer(), _src.getBytes(), kind);
+  cudaError_t err = cudaMemcpy(_dst.getBytePtr(),
+                               _src.getBytePtr(),
+                               _dst.getUnpaddedBytesInDim(0),
+                               kind);
   THROW_ON_CUDA_ERROR(err, "Failed to copy from CudaHostMemoryHeap to CudaDeviceMemory");
 }
 
 template<class Type, unsigned Dim> void copy(CudaHostMemoryHeap<Type, Dim>& _dst, const CudaArray<Type, Dim>& _src)
 {
   cudaMemcpyKind kind = cudaMemcpyDeviceToHost;
   if(Dim == 1) {
-    cudaError_t err = cudaMemcpyFromArray(_dst.getBuffer(), _src.getArray(), 0, 0, _dst.getSize()[0] * sizeof (Type), kind);
+    cudaError_t err = cudaMemcpyFromArray(_dst.getBytePtr(),
+                                          _src.getArray(),
+                                          0, 0,
+                                          _dst.getUnpaddedBytesInDim(0),
+                                          kind);
     THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
-  else if(Dim == 2) {
-    cudaError_t err = cudaMemcpy2DFromArray(_dst.getBuffer(),
-                                            _dst.getSize()[0] * sizeof (Type),
+  else if(Dim == 2)
+  {
+    cudaError_t err = cudaMemcpy2DFromArray(_dst.getBytePtr(),
+                                            _dst.getPitch(),
                                             _src.getArray(),
                                             0,
                                             0,
-                                            _dst.getSize()[0] * sizeof (Type),
-                                            _dst.getSize()[1],
+                                            _dst.getUnpaddedBytesInDim(0),
+                                            _dst.getUnpaddedBytesInDim(1),
                                             kind);
     THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
-  else if(Dim == 3) {
+  else if(Dim >= 3)
+  {
+    size_t number_of_rows = 1;
+    for( int i=1; i<Dim; i++ ) number_of_rows *= _dst.getUnitsInDim(i);
+
     cudaMemcpy3DParms p = { 0 };
     p.srcArray = const_cast<cudaArray *>(_src.getArray());
-    p.srcPtr.pitch = _src.getPitch();
-    p.srcPtr.xsize = _src.getSize()[0];
-    p.srcPtr.ysize = _src.getSize()[1];
-    p.dstPtr.ptr = (void *)_dst.getBuffer();
-    p.dstPtr.pitch = _dst.getSize()[0] * sizeof (Type);
-    p.dstPtr.xsize = _dst.getSize()[0];
-    p.dstPtr.ysize = _dst.getSize()[1];
-    p.extent.width = _dst.getSize()[0];
-    p.extent.height = _dst.getSize()[1];
-    p.extent.depth = _dst.getSize()[2];
+
```

Comment:
```
p.srcPtr.pitch
p.srcPtr.xsize
p.srcPtr.ysize
```
are not set anymore, is that normal?

Reply:
Yes. the cudaMempcy3DParms structure is a strange things. If the srcArray!=0, a cudaArray is assumed as the source. The cudaArray is a hidden structure that contains the data layout information. If srcArray==0, it is assumed that the srcPtr.ptr points to Pitched memory, and the srcPtr.pitch, srcPtr.xsize and srcPtr.ysize must be used to describe it. Everything in srcPtr is ignored if srcArray!=0, so it was redundant information before.

Reply:
Ok, good to know!